daemon/api_sideload_n_try.go

```diff
@@ -32,29 +34,72 @@ import (
 	"github.com/snapcore/snapd/client"
 	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/logger"
 	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/overlord/assertstate"
-	"github.com/snapcore/snapd/overlord/auth"
 	"github.com/snapcore/snapd/overlord/snapstate"
 	"github.com/snapcore/snapd/overlord/state"
 	"github.com/snapcore/snapd/snap"
 	"github.com/snapcore/snapd/snap/snapfile"
 )
 
-const maxReadBuflen = 1024 * 1024
+type Form struct {
+	Value      map[string][]string
+	FileHeader map[string][]*FileHeader
+}
+
+type FileHeader struct {
+	Filename string
+	TmpPath  string
+}
+
+func (f *Form) RemoveAll() {
+	for _, headers := range f.FileHeader {
+		for _, header := range headers {
+			if err := os.Remove(header.TmpPath); err != nil {
+				logger.Noticef("cannot remove temporary file: %v", err)
+			}
+		}
+	}
+}
+
+// SnapFileNameAndPath returns the original file path/name and the path to
+// where the temp file is written.
+func (f *Form) SnapFileNameAndPath() (name, path string, apiErr *apiError) {
+	if len(f.FileHeader["snap"]) == 0 {
+		return "", "", BadRequest(`cannot find "snap" file field in provided multipart/form-data payload`)
+	}
 
-func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string, user *auth.UserState) Response {
+	snapFile := f.FileHeader["snap"][0]
+	name, path = snapFile.Filename, snapFile.TmpPath
+
+	if len(f.Value["snap-path"]) > 0 {
+		name = f.Value["snap-path"][0]
```

Comment:
That would replace the path to temporary file, which we wouldn't want. `name` is the name of the source file so this replaces that (which was the original behavior). Although, I think after this point it's only used for error/summary messages

Reply:
ah, got it :-) But the naming of the variables make it rather confusing, because at first sight it looks like we are talking about the file name and path components for the same file. Maybe rename them:
- name -> targetPath
- path -> sourcePath (or tmpPath)

Reply:
I think name can also be just the filename and not the full path, depending how it's used. So maybe sourceName and tmpPath?

Reply:
I guess you mean "target" instead of "source" (unless I misunderstood the flow)?

Reply:
Source as in what it's called in the request (`srcFilename` might be clearer)